### DISCUSSION TITLE
Resampling vs Subsampling (frequential scattering)

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#721] [Q&A]
[Origin](https://github.com/kymatio/kymatio/discussions/716); split for focus. Still relevant with implems 1, 2, & 3.

## 2. Resampling vs subsampling

Since shorter padding here corresponds to a _trimming_, not subsampling, the 'observation time' per row is the same (i.e., row-to-row stays `dw`, not `2*dw`).

 1. **Resample**: keeps filters tailored for the same frequential scale `F` across all paddings
 2. **Subsample**: recalibrates filters to scales `F/2`, `F/4`, ..., or x2, x4 frequencies

Seems there's a use case for each: 

  - 1 enforces same degree of frequential invariance across all `psi2`. Disadvantages:
      - a. Suffers at highest scales: if e.g. min padded is `32` and max in `128`, then lowest frequency wavelets will target long-term structures that do not exist _and_ incur significant boundary effects
      - b. Deprioritizes sharp dynamics (less filters dedicated to high frequencies with tight temporal support)
  - 2 better exploits the "zoom" advantage by recalibrating as if less-padded is the intended input size. Disadvantages:
      - Less than user-set frequential invariance

Exact code on [padding](https://github.com/lostanlen/kymatio/blob/7d12cffd2fcebbd9b6e8aeb7ee4af573c6058c46/kymatio/scattering1d/core/timefrequency_scattering.py#L112), [resampling](https://github.com/lostanlen/kymatio/blob/7d12cffd2fcebbd9b6e8aeb7ee4af573c6058c46/kymatio/scattering1d/filter_bank.py#L756).

I'm inclined toward subsampling.

**Comment by lostanlen** [comment]
answered in https://github.com/lostanlen/kymatio/pull/1 ?

**Reply by OverLordGoldDragon** [reply]
Yes, but here I'd like to post a demo of `resample_* = False`'s advantage over `= True`, along a summary of our discussions.

**Comment by OverLordGoldDragon** [comment]
There's a problem with `resample_phi_fr = True`: shortest filter ends up doing global average pooling, and most filters deviate significantly from original's time behavior (defeating purpose of resampling):

![image](https://user-images.githubusercontent.com/16495490/115931163-b7795380-a49b-11eb-8f3c-c188bfe37157.png)

It's clear from first plot that we can downsize by up to `2^1` without significantly altering the lowpass; subsampling does not suffer from this.

This means we cannot subsample (in `_frequency_scattering`) beyond said bound _before_ `_joint_lowpass()`, defeating the entire purpose of variable padding. If the MATLAB implementation pads variably and resamples... it also suffers. Thoughts? @lostanlen

**Reply by lostanlen** [reply]
the problem only arises for large values of `F`, i.e. above one octave.
It's a difficult corner case in between global averaging pooling and small `F` (equal or lower than one octave). 
In the MATLAB implementation `F` defaults to one octave.

Global average pooling can be made numerically correct (and fast) even with a jagged array like the one we have.
Small `F` does not suffer much from this problem because the support of `phi` fits even the smallest `j2`-gram.

Large `F` is more complicated and requires more padding, i.e. `~64` coefficients in your case. Which will slow down the code but is the only solution to be numerically correct and keep some amount of frequential sensitivity. (as opposed to global average pooling)

I guess that the nicest workaround would be to account for that in the padding condition? I.e. prioritize fitting the `p*i_fr` filterbank and then consider the `j1<j2` trimming heuristic if possible

**Reply by OverLordGoldDragon** [reply]
Yes, it's what I figured, but I'm saying with the default `F` set per `2**J_fr`, there'll basically be no subsampling before `_joint_lowpass`.

I'll account for it, then.

**Comment by OverLordGoldDragon** [comment]
Turns out the `padlen(S1) > padlen(Y2)` case isn't accounted correctly with "resampling": resampling with more samples does not permit subsampling more. This forces all coefficients to double in size (i.e. the "longest input" reference now shifts from `2**J_pad_max` to `2**(J_pad_max+1)`), if we are to remain correct.

This seems too great a price to pay for sparing a bit extra boundary effects; it afflicts a fraction (`2 / (len(psi2_f) + 2)`) of coefficients, and the corrective price is _all_ joint coefficients packing more air. Instead I say we force `padlen(S1) <= padlen(Y2)` with `resample_*=True`. (The `=False` variant doesn't suffer from this).

**Reply by OverLordGoldDragon** [reply]
I've decided to scrap accounting for `padlen(S1) != padlen(Y2)` (`<` is also possible) entirely and simply force them equal; it adds much complexity without much benefit. If one insists on marginal gains, code at [this stage](https://github.com/kymatio/kymatio/pull/674/commits/d8c258f62a871fdcaa678a8781f8e721e3b9f53b) can be built off of.

_Self note_: `sc_freq.ind_start_fr[-1]` is for `J_pad_fr_max` not `J_pad_fr_fo`

**Comment by OverLordGoldDragon** [comment]
More trouble than I realized: since subsampling by `k` doesn't only `sigma *= k` but also `xi *= k`, we end up with _duplicate wavelets_, where subsampling simply "resets" parameters to an earlier combination:

<img src="https://user-images.githubusercontent.com/16495490/118333050-16227200-b51c-11eb-95aa-e0050f4069ca.png" width="480">

Red doesn't follow blue, orange does; that's the duplication, a later wavelet overlaps an earlier.

This'd be less trouble if not for exact duplication, though a complicated re-ordering would be requires to guarantee `xi[n + 1] < x[n]`. My original motivation for subsampling only concerned reducing temporal width; to this end, I propose doing exactly that. Instead of subsampling by `k`, just do `sigma *= k`, and call it `recalibrate=True` instead of `resample=False`:

![image](https://user-images.githubusercontent.com/16495490/118334644-0c4e3e00-b51f-11eb-8acf-c1aadf5ee8f6.png)

example wavelets (red, cyan above):

![image](https://user-images.githubusercontent.com/16495490/118335272-57b51c00-b520-11eb-9bf0-8dc7be98a3c1.png)

**Reply by OverLordGoldDragon** [reply]
After adjusting so that greater sigma doesn't reduce `j`:

![image](https://user-images.githubusercontent.com/16495490/118339159-1412e000-b529-11eb-808c-d048b942f3f1.png)

This is no longer CQT nor piecewise-QT, it's variable Q for some filters and CQT for others

**Reply by OverLordGoldDragon** [reply]
I've confused the definition of 'CQT', one can simply increase initial sigma (done via `r_psi`) and that carries to the rest, while keeping center frequencies and number of wavelets the same:

![image](https://user-images.githubusercontent.com/16495490/118383183-be176880-b60c-11eb-8d83-26a6ab36631b.png)

**Reply by OverLordGoldDragon** [reply]
With `Q` already low for `psi_fr`, multiplying `sigma` really messes with analyticity and temporal envelope:

![image](https://user-images.githubusercontent.com/16495490/118405873-d332dc80-b68a-11eb-8b6b-57a84abd130a.png)

This solution narrows time support of _all_ wavelets just because _some_ are too wide. A more efficient approach is to target the offenders: STFT. Set target `sigma_min` and force all wavelets to `>=`, effectively raising the portion of "intermediate" filters, which is cleaner than variably scaling sigma for each filter.

![image](https://user-images.githubusercontent.com/16495490/118406366-d4fd9f80-b68c-11eb-9e20-e2fb26c14083.png)

**Reply by OverLordGoldDragon** [reply]
I conclude it's impossible to get well-behaved wavelets that satisfy all of the following: 1) different input sizes 2) same center frequencies 3) same total number of wavelets 4) larger `sigma_min`. For one, `xi_min` must increase to avoid center frequency duplication due to lower filter length.

New idea, "linear Q transform": keep `sigma_max` same but increase `sigma_min`, `xi_min`, and `xi_max`, and distribute them all exponentially. This breaks the constant ratio between center frequency and bandwidth (instead making it linear), but makes wavelets well-behaved and doesn't require intricate handling for varying `J`, `Q`.

![image](https://user-images.githubusercontent.com/16495490/118414835-e60fd600-b6b7-11eb-9213-4c66768bfcf4.png)

**Reply by lostanlen** [reply]
i don't see why we'd need 3) same total number of wavelets

**Reply by OverLordGoldDragon** [reply]
Only for 3D+. But there's other advantages; I realized there is dropping of wavelets happening after all, it's forced by shorter filter length which limits maximum scale. Difference is, drop = use fewer existing filters, more sparse; recalibrate = redistribute existing filters, finer distribution for new max scale:

 - Greater `xi_max` -> greater discriminability of small FDTS
 - More `xi` between higher `xi_min` and new `xi_max` -> more 2D wavelet orientations in directions relevant to input length
 - Greater `sigma` for all (but last) wavelets -> more appropriate lesser width for lesser input length

`xi` discrepancy may be bad news for 4D convs, but 3D with one separable wavelet along channels permits independent treatment.

**Reply by OverLordGoldDragon** [reply]
Previous filterbanks but with resampling of different parameters at original length, for perspective:

![image](https://user-images.githubusercontent.com/16495490/118537803-730d6a80-b75e-11eb-94ff-e1364397fcf0.png)

**Reply by OverLordGoldDragon** [reply]
One way to do variable `n1_fr` per `n2` is to keep all parameters the same but _drop_ those whose sigma are beneath threshold, namely drop `<= sigma_min * 2**downsampling_factor`. Included full length for reference:

![image](https://user-images.githubusercontent.com/16495490/118539739-d4ced400-b760-11eb-978d-451535a13429.png)

Is this what you had in mind? @lostanlen

**Reply by lostanlen** [reply]
@OverLordGoldDragon yes that's exactly what i had i mind! and also my preferred option because it keeps the `n1_fr`'s xis  (when expressed in radians or cycles per octaves) aligned across different `n2`
